File: zx_view_widget/src/zxRender.py
import base64
import io
import json
from fractions import Fraction

import matplotlib
import matplotlib.pyplot as plt
import pyzx as zx
from pyzx.graph import Graph
from pyzx.graph.base import BaseGraph
from pyzx.utils import EdgeType, VertexType

matplotlib.use("Agg")


type ZXLeanGraph = BaseGraph[int, int]


def leanzx_to_pyzx(data) -> ZXLeanGraph:
    """Convert LeanZX JSON to a pyzx Graph."""
    g: ZXLeanGraph = Graph()
    nodes = data.get("nodes", [])
    edges = data.get("edges", [])

    inputs: list[int] = []
    outputs: list[int] = []

    for node in nodes:
        nid = node["id"]
        ntype = node["type"]

        # Manually adding vertex and setting type/row/qubit so that we can set
        #   the vertex indices manually, so that they align with the node IDs
        #   in PyZX
        # Essentially unrolling the g.add_vertex function
        g.add_vertex_indexed(nid)

        if ntype == "input":
            g.set_type(nid, VertexType.BOUNDARY)
            g.set_row(nid, 0)
            g.set_qubit(nid, node.get("ioId", 0))
            inputs.append(nid)
        elif ntype == "output":
            g.set_type(nid, VertexType.BOUNDARY)
            g.set_row(nid, -1)
            g.set_qubit(nid, node.get("ioId", 0))
            outputs.append(nid)
        elif ntype == "spider":
            color = node.get("color", "Z")
            ty = VertexType.Z if color == "Z" else VertexType.X
            phase_str = node.get("phase", "0")
            phase = _parse_phase(phase_str)
            g.set_type(nid, ty)
            g.set_phase(nid, phase)
        elif ntype == "hadamard":
            phase_str = node.get("phase", "0")
            phase = _parse_phase(phase_str)
            g.set_type(nid, VertexType.H_BOX)
            g.set_phase(nid, phase)
        else:
            g.set_type(nid, VertexType.BOUNDARY)

    for edge in edges:
        g.add_edge((edge["src"], edge["tgt"]), edgetype=EdgeType.SIMPLE)

    g.set_inputs(tuple(inputs))
    g.set_outputs(tuple(outputs))

    # Auto-layout spider positions
    _auto_layout(g)

    return g


def _parse_phase(s):
    """Parse a phase string like '0', '1', '1/2' into a Fraction."""
    s = s.strip()
    if "/" in s:
        return Fraction(s)
    return Fraction(int(s))


def _auto_layout(g: ZXLeanGraph):
    """Simple left-to-right layout based on graph distance from inputs."""
    inputs = g.inputs()
    outputs = g.outputs()
    if not inputs:
        return

    # BFS from inputs to assign rows to interior vertices
    visited = {}
    queue = list(inputs)
    for v in queue:
        visited[v] = 0

    while queue:
        current = queue.pop(0)
        for neighbor in g.neighbors(current):
            if neighbor not in visited:
                visited[neighbor] = visited[current] + 1
                queue.append(neighbor)

    # Set rows for non-boundary vertices
    max_depth = max(visited.values(), default=1)
    for v, depth in visited.items():
        if v not in inputs and v not in outputs:
            g.set_row(v, depth)

    # Set output row to max_depth
    for v in outputs:
        g.set_row(v, max_depth)

    # Assign qubit indices to interior vertices, avoiding collisions
    # First, record which qubit slots are already taken per row (by inputs/outputs)
    row_taken: dict[int, set[int]] = {}
    for v in inputs:
        r = int(g.row(v))
        row_taken.setdefault(r, set()).add(int(g.qubit(v)))
    for v in outputs:
        r = int(g.row(v))
        row_taken.setdefault(r, set()).add(int(g.qubit(v)))

    row_counts: dict[int, int] = {}
    for v in g.vertices():
        if v in inputs or v in outputs:
            continue
        r = int(g.row(v))
        count = row_counts.get(r, 0)
        taken = row_taken.get(r, set())
        # Skip slots occupied by boundary nodes
        while count in taken:
            count += 1
        g.set_qubit(v, count)
        row_counts[r] = count + 1


def render_to_base64(g: ZXLeanGraph) -> str:
    """Render a pyzx graph to a base64-encoded PNG string."""
    # Scale figsize based on graph dimensions
    rows = set(g.row(v) for v in g.vertices())
    qubits = set(g.qubit(v) for v in g.vertices())
    width = max(len(rows) * 1.5, 6)
    height = max(len(qubits) * 1.2, 2)
    fig = zx.draw_matplotlib(g, labels=True, figsize=(width, height))
    buf = io.BytesIO()
    fig.savefig(buf, format="png", bbox_inches="tight", dpi=150)
    plt.close(fig)
    buf.seek(0)
    return base64.b64encode(buf.read()).decode("ascii")


def render(diagram_json: str) -> str:
    """Convert a ZXDiagram JSON string to a base64-encoded PNG via pyzx."""
    data = json.loads(diagram_json)
    g = leanzx_to_pyzx(data)
    image = render_to_base64(g)
    return image

File: zx_view_widget/src/zxRender.ts
// TypeScript port of the former zxRender.py: builds a D3-ready graph object
// from the Lean diagram JSON, replacing pyzx + Pyodide.

export interface DiagramNode {
  id: number
  type: 'spider' | 'input' | 'output' | 'hadamard'
  color?: 'Z' | 'X'
  phase?: string
  ioId?: number
}

export interface DiagramEdge {
  src: number
  tgt: number
}

export interface DiagramData {
  nodes: DiagramNode[]
  edges: DiagramEdge[]
}

export interface GraphNode {
  name: string
  x: number
  y: number
  t: number
  phase: string
  ground: boolean
  vdata: [string, unknown][]
}

export interface GraphLink {
  source: string
  target: string
  t: number
  index: number
  num_parallel: number
}

export interface GraphData {
  nodes: GraphNode[]
  links: GraphLink[]
  pauli_web: never[]
}

export interface RenderData {
  graph: GraphData
  width: number
  height: number
  scale: number
  node_size: number
  colors: Record<string, string>
}

const VertexType = { BOUNDARY: 0, Z: 1, X: 2, H_BOX: 3 } as const
const EdgeType = { SIMPLE: 1 } as const

// pyzx.utils.original_colors
const COLORS: Record<string, string> = {
  edge: '#000000',
  Hedge: '#0088ff',
  Xedge: '#999999',
  boundary: '#000000',
  X: '#ff8888',
  Y: '#aabbff',
  Z: '#ccffcc',
  H: '#ffff66',
  W: '#000000',
  Zalt: '#ccffcc',
  Walt: '#000000',
  Xdark: '#ff8888',
  Ydark: '#aabbff',
  Zdark: '#99dd99',
}

function gcd(a: number, b: number): number {
  let x = Math.abs(a)
  let y = Math.abs(b)
  while (y) {
    const t = y
    y = x % y
    x = t
  }
  return x || 1
}

interface Phase { num: number; den: number }

function parsePhase(s: string | undefined): Phase {
  const str = (s ?? '0').trim()
  let num: number
  let den: number
  if (str.includes('/')) {
    const [n, d] = str.split('/')
    num = Number.parseInt(n, 10)
    den = Number.parseInt(d, 10)
  } else {
    num = Number.parseInt(str, 10)
    den = 1
  }
  if (den < 0) { num = -num; den = -den }
  const g = gcd(num, den)
  return { num: num / g, den: den / g }
}

// pyzx.utils.phase_fraction_to_s for spiders/H-boxes (poly_with_pi irrelevant
// here because Lean phases are always rational).
function phaseToString(p: Phase, t: number): string {
  const isHBox = t === VertexType.H_BOX
  if (p.num === 0 && !isHBox) return ''
  if (p.num === p.den && isHBox) return ''
  if (p.num === 0) return '0'
  const ns = p.num === 1 ? '' : String(p.num)
  const ds = p.den === 1 ? '' : `/${String(p.den)}`
  return `${ns}π${ds}`
}

interface InternalNode {
  id: number
  t: number
  row: number | null
  qubit: number | null
  phaseStr: string
  isInput: boolean
  isOutput: boolean
}

function buildNodes(diagram: DiagramData): Map<number, InternalNode> {
  const out = new Map<number, InternalNode>()
  for (const n of diagram.nodes) {
    let t: number
    let row: number | null = null
    let qubit: number | null = null
    let phaseStr = ''
    let isInput = false
    let isOutput = false

    if (n.type === 'input') {
      t = VertexType.BOUNDARY
      row = 0
      qubit = n.ioId ?? 0
      isInput = true
    } else if (n.type === 'output') {
      t = VertexType.BOUNDARY
      row = -1
      qubit = n.ioId ?? 0
      isOutput = true
    } else if (n.type === 'spider') {
      t = n.color === 'X' ? VertexType.X : VertexType.Z
      phaseStr = phaseToString(parsePhase(n.phase), t)
    } else if (n.type === 'hadamard') {
      t = VertexType.H_BOX
      phaseStr = phaseToString(parsePhase(n.phase ?? '1'), t)
    } else {
      t = VertexType.BOUNDARY
    }
    out.set(n.id, { id: n.id, t, row, qubit, phaseStr, isInput, isOutput })
  }
  return out
}

function autoLayout(nodes: Map<number, InternalNode>, edges: DiagramEdge[]): void {
  const inputs: number[] = []
  const outputs: number[] = []
  for (const n of nodes.values()) {
    if (n.isInput) inputs.push(n.id)
    if (n.isOutput) outputs.push(n.id)
  }
  if (inputs.length === 0) return

  const adj = new Map<number, number[]>()
  for (const id of nodes.keys()) adj.set(id, [])
  for (const e of edges) {
    adj.get(e.src)?.push(e.tgt)
    adj.get(e.tgt)?.push(e.src)
  }

  // BFS depth from inputs
  const depth = new Map<number, number>()
  const queue: number[] = []
  for (const id of inputs) {
    depth.set(id, 0)
    queue.push(id)
  }
  while (queue.length) {
    const cur = queue.shift() as number
    const curDepth = depth.get(cur) ?? 0
    for (const nb of adj.get(cur) ?? []) {
      if (!depth.has(nb)) {
        depth.set(nb, curDepth + 1)
        queue.push(nb)
      }
    }
  }

  let maxDepth = 1
  for (const d of depth.values()) if (d > maxDepth) maxDepth = d

  // Assign rows: interior non-H-box vertices get BFS depth; outputs get maxDepth.
  // H-boxes get no row (positioned by auto_hbox in the viewer).
  for (const [id, d] of depth) {
    const n = nodes.get(id)
    if (!n) continue
    if (!n.isInput && !n.isOutput && n.t !== VertexType.H_BOX) {
      n.row = d
    }
  }
  for (const id of outputs) {
    const n = nodes.get(id)
    if (n) n.row = maxDepth
  }

  // Qubit assignment: interior non-H-box vertices fill the lowest free slot per
  // row, skipping qubit indices already claimed by boundary nodes on that row.
  const rowTaken = new Map<number, Set<number>>()
  const claim = (row: number, q: number) => {
    let s = rowTaken.get(row)
    if (!s) {
      s = new Set()
      rowTaken.set(row, s)
    }
    s.add(q)
  }
  for (const id of [...inputs, ...outputs]) {
    const n = nodes.get(id)
    if (n && n.row !== null && n.qubit !== null) claim(n.row, n.qubit)
  }

  const rowCounts = new Map<number, number>()
  // Iterate in id order (matches pyzx's GraphS.vertices() ordering).
  const sortedIds = [...nodes.keys()].sort((a, b) => a - b)
  for (const id of sortedIds) {
    const n = nodes.get(id)
    if (!n) continue
    if (n.isInput || n.isOutput || n.t === VertexType.H_BOX) continue
    if (n.row === null) continue
    let count = rowCounts.get(n.row) ?? 0
    const taken = rowTaken.get(n.row) ?? new Set<number>()
    while (taken.has(count)) count++
    n.qubit = count
    rowCounts.set(n.row, count + 1)
  }
}

function pairKey(a: number, b: number): string {
  return a < b ? `${a}|${b}` : `${b}|${a}`
}

export function render(diagram: DiagramData): RenderData {
  const nodes = buildNodes(diagram)
  autoLayout(nodes, diagram.edges)

  // Skip H-boxes (no row/qubit) when computing the bounds.
  let minrow = Number.POSITIVE_INFINITY
  let maxrow = Number.NEGATIVE_INFINITY
  let minqub = Number.POSITIVE_INFINITY
  let maxqub = Number.NEGATIVE_INFINITY
  for (const n of nodes.values()) {
    if (n.row === null || n.qubit === null) continue
    if (n.row < minrow) minrow = n.row
    if (n.row > maxrow) maxrow = n.row
    if (n.qubit < minqub) minqub = n.qubit
    if (n.qubit > maxqub) maxqub = n.qubit
  }
  if (!Number.isFinite(minrow)) {
    minrow = 0; maxrow = 0; minqub = 0; maxqub = 0
  }

  let scale = 800 / (maxrow - minrow + 2)
  if (scale > 50) scale = 50
  if (scale < 20) scale = 20

  const width = (maxrow - minrow + 2) * scale
  const height = (maxqub - minqub + 3) * scale
  const node_size = Math.max(0.2 * scale, 2)

  // Emit nodes in id order. H-boxes without a row land at the top-left
  // placeholder coordinate; the viewer's update_hboxes() repositions them
  // before the first paint.
  const sortedIds = [...nodes.keys()].sort((a, b) => a - b)
  const outNodes: GraphNode[] = sortedIds.map(id => {
    const n = nodes.get(id) as InternalNode
    const row = n.row ?? 0
    const qubit = n.qubit ?? 0
    return {
      name: String(n.id),
      x: (row - minrow + 1) * scale,
      y: (qubit - minqub + 2) * scale,
      t: n.t,
      phase: n.phaseStr,
      ground: false,
      vdata: [],
    }
  })

  // Parallel edge metadata: group by unordered endpoint pair so (a,b) and (b,a)
  // count as parallel. Mirrors pyzx.drawing.graph_json's index/num_parallel.
  const counts = new Map<string, number>()
  const linkKeys: string[] = []
  const links: GraphLink[] = diagram.edges.map(e => {
    const k = pairKey(e.src, e.tgt)
    const i = counts.get(k) ?? 0
    counts.set(k, i + 1)
    linkKeys.push(k)
    return {
      source: String(e.src),
      target: String(e.tgt),
      t: EdgeType.SIMPLE,
      index: i,
      num_parallel: 0,
    }
  })
  for (let i = 0; i < links.length; i++) {
    links[i].num_parallel = counts.get(linkKeys[i]) ?? 1
  }

  return {
    graph: { nodes: outNodes, links, pauli_web: [] },
    width,
    height,
    scale,
    node_size,
    colors: COLORS,
  }
}
